test/integration/auth_test.go

```diff
@@ -725,6 +725,89 @@ func TestUnknownUserIsUnauthorized(t *testing.T) {
 	}
 }
 
+type actAsAuthorizer struct{}
+
+// alice can't act as anyone and bob can't do anything but act-as someone
+func (actAsAuthorizer) Authorize(a authorizer.Attributes) error {
+	if a.GetUserName() == "alice" && a.GetVerb() != "act-as" {
+		return nil
+	}
+	if a.GetUserName() == "bob" && a.GetVerb() == "act-as" {
+		return nil
+	}
+	return errors.New("I can't allow that.  Go ask alice.")
+}
+
+func TestActAsIsForbidden(t *testing.T) {
+	framework.DeleteAllEtcdKeys()
+
+	// This file has alice and bob in it.
+	var m *master.Master
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		m.Handler.ServeHTTP(w, req)
+	}))
+	// TODO: Uncomment when fix #19254
+	// defer s.Close()
+
+	masterConfig := framework.NewIntegrationTestMasterConfig()
+	masterConfig.Authenticator = getTestTokenAuth()
+	masterConfig.Authorizer = actAsAuthorizer{}
+	m, err := master.New(masterConfig)
+	if err != nil {
+		t.Fatalf("error in bringing up the master: %v", err)
+	}
+
+	transport := http.DefaultTransport
+
+	for _, r := range getTestRequests() {
+		token := BobToken
+		bodyBytes := bytes.NewReader([]byte(r.body))
+		req, err := http.NewRequest(r.verb, s.URL+r.URL, bodyBytes)
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+
+		func() {
+			resp, err := transport.RoundTrip(req)
+			defer resp.Body.Close()
+			if err != nil {
+				t.Logf("case %v", r)
+				t.Fatalf("unexpected error: %v", err)
+			}
+			// Expect all of bob's actions to return Forbidden
+			if resp.StatusCode != http.StatusForbidden {
+				t.Logf("case %v", r)
+				t.Errorf("Expected not status Forbidden, but got %s", resp.Status)
+			}
+		}()
+	}
+
+	for _, r := range getTestRequests() {
+		token := BobToken
+		bodyBytes := bytes.NewReader([]byte(r.body))
+		req, err := http.NewRequest(r.verb, s.URL+r.URL, bodyBytes)
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
+		req.Header.Set("Act-As", "alice")
```

Comment:
> Does the system just trust Bob's determination that the end user was really alice?

Bob's power to do that is controlled based on whether the API server authorizer says that Bob has the power to "act-as", the resource "users", with name "Alice".  For this test, its set up with this mock here: https://github.com/kubernetes/kubernetes/pull/23549/files/85844c419933c4a507225d52bfa542f8922685e1#diff-8754bfba92df8f58b6e09a4200975a90R735 .